keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,766 @@
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a generic feature to protect objects from deletion while it is marked as in-use.
+
+## Motivation
+
+Currently, "a generic mechanism that prevents a resource from being deleted when it is in-use" doesn't exist.
+Instead, for such a use case, each controller, like [pv-protection](https://github.com/kubernetes/enhancements/issues/499) and [pvc-protection](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md), implements its own logic to protect used objects from being deleted while in-use.
+These controllers use [Finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/) to block deletion.
+Finalizers block the _completion_ of a delete operation, but they do not prevent the deletion from starting.
+Once started, a delete _will_ complete and can not be aborted.
+
+However, Finalizer may not be appropriate for some resource protection use cases, because it doesn't prevent other controllers from executing their pre-deletion actions.
+The order of execution of pre-deletion actions across finalizers is not defined.
+As a result, other controllers will execute their pre-deletion actions while a finalizer for protection still exists.
+
+This KEP aims to provide a generic mechanism for protection that solves the above issues with Finalizers, because such use cases are also found for protecting secrets and configmaps in [KEP-2639](https://github.com/kubernetes/enhancements/pull/2640).
+
+Potential use cases are:
```

Comment:
Based on these stated use cases, I don't think Liens is the right solution to this problem.

All of these are examples of object dependencies. Some of them are parent-child relationships, and others are sibling relationships. But they're all dependencies.

Here's some of my reasoning...

### Liens are hard to understand & debug
Also, my experience using Liens on other platforms (like GCP), I have always found them to be very hard to understand and debug for users. A dependency, on the other hand, is a well understood concept intuitively graspable by all technical users. A blocked action from a dependency can tell you exactly what dependencies block the action, which can be used to describe _why_ the aciton was blocked. A Lien on the other hand does not describe why, only that the action is prohibited. To learn why you have to dig around to figure out what made the Lien and ask them why.

### Liens are negative by connotation
Liens are something the IRS takes out on your house and banks take on your car. They're almost always interpreted as negative, even if they do facilitate lending and repayment. I've never had a user ask me to implement liens in a software. What they ask for is policies, constraints, and dependencies.

### Liens are an intermediate data format
Prevention of actions is easily implementable with a webhook. So enforcement of Liens is simple. The complex part is defining them. And if you define them as Liens, you're effectively defining an intermediary data format between intent and behavior. Why not just describe the intent?

### Finalization and Initialization parity
We already have a problem that the behavior from Finalizers are not replicated by Initializers. I'm loath to put another Finalization-like behavior into the API without thinking through the inverse Initialization behaviors that we might not be able to achieve if we choose the Finalization interface first. If you think about liens as "deletion ordering requirements" it becomes obvious that there are also "creation ordering requirements" in the opposite order. 

### Apply ordering
If you try to implement creation and deletion ordering (as I have in [kpt dependency ordering](https://kpt.dev/book/06-deploying-packages/03-handling-dependencies)), it becomes apparent that you might also need  "apply ordering requirements" to apply updates in the right order across multiple resource objects. In fact, kubectl already does this when it applies a directory, by applying the files alphanumerically and top to bottom. This allows users to specify the order that objects get applied and deleted, when done so all together with a single kubectl command.

### Object References
Many of the uses cases here already have ways to reference the object they depend on (dependent references dependency), for example PersistentVolumeClaim `spec.storageClassName`. Sometimes, the reference is inverted (dependency references to dependent), for example PersistentVolumeClaim `spec.claimRef`, which is used for accounting. If we could standardize these types of references, we could define the dependency generically on the CRD as being a quality of the resource, and not something specific to each object. And of course, what we can do for CRDs in yaml, we can do in code for core objects and extensions.

So in summary, instead of Liens, I would rather pursue object references and dependencies. This way we can then implement not just deletion prevention, but also client-side create/apply/delete ordering. Granted, that's a more ambitious project, but it's also a better more holistic solution.

/cc @bgrant0607 @thockin @apelisse

Reply:
Should we have generalized referential integrity in Kubernetes?  Maybe.  It would be a VERY different system and I just don't see a path from here to there in any realistic timeframe.

I mean, it's "just software".  We could do anything IN THEORY, but I don't see this happening any time soon (and I can think of a LOT of things that I'd spend time on first).

I hate the idea of calcification of systems, but I also hate not doing X because Y would be better, and then not doing Y.  If someone wants to explore how to build the full graph of resources into the apiserver (where each edge carries policy), and figure out how that changes the sytem overall (compatibly, of course), and how it scales at the API and storage layer, I am more than willing to help chart the course.  I have a lot of ideas and even more questions.

But I don't think we should derail useful incremental work in the mean time.

Reply:
Apiserver has no control over the order of multi-object apply/creates. We took initializers out because it was too hard to make the entirety of the system deal with partially constructed objects, so I don't see this changing.

Additionally, we are almost certainly never going to have transactions.

We actually already have dependencies -- ownerReferences. But they are for auto-tear-down, not for preventing deletion.

I do agree with the point about liens being hard to debug, and it's worth thought about how we can make that understandable. I don't think liens should have to be inherently more hard to understand than a dependency, which can also be inscrutable (why does X depend on this random custom resource??). Additionally you can implement a dependency system using liens, if you want to.

API features are often "some assembly required" because that way you can solve more use cases with fewer mechanisms (moving complexity for specific use cases into plugins/controllers).

Reply:
To be clear, I could imagine a world where:

* Pointer-like fields are tagged "this is a pointer (to a fixed resource or a dynamic one)"
* Pointer fields carry policy like
  - "inhibit deletion of the referent"; or
  - "delete me when the referent is deleted"; or
  - "inhibit deletion of me as long as the referent exists"; or
  - "fail creation if the referent does not exist"
* "The system" enforces those edges

What I don't know is, well, everything else.  How will it scale?  How will it impact API complexity?  How do I prevent errant referrers?  How do I force-override these?  what fun new consistency issues arise?  

It's just a LONG way from here.  It sounds like a fun project, though.

Reply:
> I hate the idea of calcification of systems, but I also hate not doing X because Y would be better, and then not doing Y... But I don't think we should derail useful incremental work in the mean time.

In a world where reverse incompatible API changes are possible with major version bumps, I agree that preventing incremental change is bad. But I think the Kubernetes reverse compatible guarantees have calcified the system in such a way that making a bad API change (one that prevents future iteration by not planning for it) is worse than not making that change.

>  If someone wants to explore how to build the full graph of resources into the apiserver (where each edge carries policy), and figure out how that changes the sytem overall (compatibly, of course), and how it scales at the API and storage layer, I am more than willing to help chart the course. I have a lot of ideas and even more questions.

We're already built a system into kpt that builds a dependency graph and handles ordering of applies and deletes. It still needs some work before I would recommend it for general use (I plan to make it do parallel apply, instead of serial apply), but it's not particularly _hard_ to build. It's "just software" as you say.

The key constraint in what kpt does today is that it's limited to a single "package" or set of objects managed together. Managing incoming dependencies from any object on the cluster would have to be server-side, not client-side. But that's something that could be enforced by a webhook, even if the webhook can't perform the ordering of applies/deletes. A webhook would simply reject out-of-order requests with an error message that explains the reasoning. As long as a client can interpret the same dependencies, the client can perform ordering of operations for the set of objects it cares about.

> Apiserver has no control over the order of multi-object apply/creates. We took initializers out because it was too hard to make the entirety of the system deal with partially constructed objects, so I don't see this changing.

Agreed. I don't think the server can or should re-order apply operations. To do so would require batch operations and/or transactions, which, as you say, is unlikely.

However, it's entirely reasonable for the server to simply to enforcement to prevent unordered operations (deletion ordering & creation ordering). Then it's up to the client to handle ordering operations so that they satisfy those requirements.

Reply:
All that stuff can be built via webhooks if the system offered the lien feature. It is tricky to get it right. I want people to experiment and build it. But it's very slow and hard to do so in-tree. Plugins don't have compatibility problems until people actually start using them, so they're much faster to develop.

I see liens as enabling that behavior, not as competing with it. (You can't build this today because there's no way to prevent a deletion.)

Reply:
> webhooks can prevent a delete request, right?

Good point. In that case I'd like to see an out-of-tree prototype built before we do something in-tree?

Reply:
A few other points:

### Self Liens
It's nonsensical for a property owner to take a lien on their own property. Liens are always taken (in real life) by a 3rd party who needs to legally prevent the owner from performing a sale or transfer. So it doesn't make sense to use a "lien" to allow an owner to prevent "accidental deletion" of their own objects. Liens ARE dependencies. So you might as well say "bank X requires payment Y before property Z is allowed to be sold. And then you can invert it into a construction policy: "city X requires permit Y before property Z can be built." They're both dependencies, but only one of them is a Lien.

## Deletion protection
Imagine the "don't delete this database" use case in GitOps. In interactive modes it might trigger an "are you sure?" prompt or require a "--force" flag. But in GitOps, neither of those are possible, because not only are the manifests declarative but the operations performed by the client are also declarative. The object either exists in the repo directory and is created/updated or it was removed and is deleted. How do you remove a Lien in this case? Is it really any better than just removing the object?

## Defining operation ordering
Say we add Liens. How does a user now apply their objects in the right order? With kubectl/kustomize, they have to alphabetize their files or arrange objects top to bottom in a file. This affects both create, update, and deletion ordering. With kpt you get some resource-based sorting for free (covers most of these uses cases) or you can manually define dependencies. But do you really want the user to need to manually order their applies? What if the server-side enforcement and apply/delete ordering both used the same interface? Wouldn't that be better?

Reply:
In this system, typically the lienholder wouldn't be the object itself.

I think close to the whole point of liens is to break gitops workflows when they're about to clobber something important.

I don't see why liens affect creation order.

Reply:
> All that stuff can be built via webhooks if the system offered the lien feature.

If liens only define deletion prevention, then no, the API won't be sufficient to also define creation and update ordering. A "lien" exclusively prevents deletion (aka sale or transfer). If we want it to also specify create & apply ordering, we can't call it a lien. Perhaps this is quibbling over naming, but I think the naming is important to the meaning the usage specifies.

Reply:
I continue to not see why you think a lien has or should have an effect on creation or update ordering?

Reply:
> I think close to the whole point of liens is to break gitops workflows when they're about to clobber something important.

There's two distinct use cases being described. One is that the object is in-use. The other is that the object is "important". Personally, I think if we define dependencies to satisfy the in-use cases (e.g. Deployment (backend) using a StatefulSet (Database)) then most of the other use cases evaporate, or can be defined in terms of a dependency.

> I don't see why liens affect creation order.

"Liens" themselves do not. However, if we define an interface for deletion order, we should also define an interface for creation order, otherwise we end up with the same asymmetry we have with Initializers/Finalizers. There isn't an inverse word for Lien (AFAIK) that we could add later to handle create ordering. But "Dependencies" works both ways.

Reply:
> To be clear, I could imagine a world where:
> 
> * Pointer-like fields are tagged "this is a pointer (to a fixed resource or a dynamic one)"
> * Pointer fields carry policy like
>   
>   * "inhibit deletion of the referent"; or
>   * "delete me when the referent is deleted"; or
>   * "inhibit deletion of me as long as the referent exists"; or
>   * "fail creation if the referent does not exist"; or
>   * "fail updates that break this reference"; or
>   * ... ad infinitum
> * "The system" enforces those edges

Dependency references should be able to handle most of these:
- Disallow deletion of that object before this object is deleted (deletion ordering)
- Disallow creation of this object before that object exists (creation ordering)
- Disallow update of this object before that new object exists (update ordering)

The one left out is: 
- Delete me when the referent is deleted (garbage collection)
But that's already handled by resource ownership today.

> How will it scale? How will it impact API complexity? How do I prevent errant referrers? How do I force-override these? what fun new consistency issues arise?

Good questions. I don't have immediate answers. I didn't come with a full design spec'd out. Only a warning to not implement half the API in a way that prevents adding the rest later.

Reply:
> , if we define an interface for deletion order, we should also define an interface for creation order, otherwise we end up with the same asymmetry we have with Initializers/Finalizers. There isn't an inverse word for Lien (AFAIK) that we could add later to handle create ordering. But "Dependencies" works both ways.

I'm not following this logic at all, I don't see that symmetry as existing *at all* in a world where clients do the create/update ordering and it can't be affected by the server. Even if it was a thing, I don't see how it would affect liens, IMO it is good that the name can only mean the one thing and we'd need a different name if we wanted it to do a different thing.

Reply:
> I think if we define dependencies to satisfy the in-use cases (e.g. Deployment (backend) using a StatefulSet (Database)) then most of the other use cases evaporate, or can be defined in terms of a dependency.

I don't see it that way, I would prefer a very simple thing that works for all use cases than a complicated thing that we think might work for most use cases.

Reply:
> I don't see it that way, I would prefer a very simple thing that works for all use cases than a complicated thing that we think might work for most use cases.

For most other projects I agree with you. But as I said earlier, I think Kubernetes' reverse compatibility guarantees makes it really hard to iterate on simple things, because once they're GA they can't be modified into a more complex thing without being reverse compatible. Naming is one big thing you can't take back, and naming informs the intent and usability here.

Reply:
I saw this discussion and I thought of:

> [Cloud Native architectures] enable **loosely coupled** systems that are resilient, manageable, and observable

(my emphasis, quoting from the [cloud native definition](https://github.com/cncf/toc/blob/main/DEFINITION.md))

Reply:
I talked with Karl to understand better his use cases and I promised to at least imagine how a dependency-based solution for this would look.